File: clients/android/app/src/main/kotlin/com/lextures/android/core/lms/PeerReviewLogic.kt
package com.lextures.android.core.lms

import com.lextures.android.R

object PeerReviewLogic {
    fun targetLabel(allocation: PeerReviewAllocation): String? =
        allocation.targetLabel?.takeIf { it.isNotBlank() }

    fun isComplete(allocation: PeerReviewAllocation): Boolean =
        allocation.status == PeerReviewAllocationStatus.submitted

    fun pending(allocations: List<PeerReviewAllocation>): List<PeerReviewAllocation> =
        allocations.filterNot(::isComplete)

    fun completedCount(allocations: List<PeerReviewAllocation>): Int =
        allocations.count(::isComplete)

    fun rubricTotal(rubric: RubricDefinition, scores: Map<String, Double>): Double =
        rubric.criteria.sumOf { scores[it.id] ?: 0.0 }

    fun rubricGradedCount(rubric: RubricDefinition, scores: Map<String, Double>): Int =
        rubric.criteria.count { scores.containsKey(it.id) }

    fun rubricScoresComplete(rubric: RubricDefinition, scores: Map<String, Double>): Boolean {
        if (rubric.criteria.isEmpty()) return true
        return rubricGradedCount(rubric, scores) == rubric.criteria.size
    }

    fun statusLabelRes(status: PeerReviewAllocationStatus): Int = when (status) {
        PeerReviewAllocationStatus.assigned -> R.string.mobile_peerReview_statusAssigned
        PeerReviewAllocationStatus.in_progress -> R.string.mobile_peerReview_statusInProgress
        PeerReviewAllocationStatus.submitted -> R.string.mobile_peerReview_statusSubmitted
        PeerReviewAllocationStatus.expired -> R.string.mobile_peerReview_statusExpired
    }

    fun cacheKeyAssigned(): String = "peer-review:assigned"

    fun cacheKeyAllocation(id: String): String = "peer-review:allocation:$id"

    fun cacheKeyReceived(courseCode: String, assignmentId: String): String =
        "peer-review:received:$courseCode:$assignmentId"
}

File: clients/ios/Lextures/Core/LMS/PeerReviewLogic.swift
import Foundation

enum PeerReviewLogic {
    static func targetLabel(_ allocation: PeerReviewAllocation) -> String {
        allocation.targetLabel?.trimmingCharacters(in: .whitespacesAndNewlines).nonEmpty
            ?? L.text("mobile.peerReview.anonymousPeer")
    }

    static func isComplete(_ allocation: PeerReviewAllocation) -> Bool {
        allocation.status == .submitted
    }

    static func pending(_ allocations: [PeerReviewAllocation]) -> [PeerReviewAllocation] {
        allocations.filter { !isComplete($0) }
    }

    static func completedCount(_ allocations: [PeerReviewAllocation]) -> Int {
        allocations.filter(isComplete).count
    }

    static func rubricTotal(_ rubric: RubricDefinition, scores: [String: Double]) -> Double {
        rubric.criteria.reduce(0) { partial, criterion in
            partial + (scores[criterion.id] ?? 0)
        }
    }

    static func rubricGradedCount(_ rubric: RubricDefinition, scores: [String: Double]) -> Int {
        rubric.criteria.filter { scores[$0.id] != nil }.count
    }

    static func rubricScoresComplete(_ rubric: RubricDefinition, scores: [String: Double]) -> Bool {
        guard !rubric.criteria.isEmpty else { return true }
        return rubricGradedCount(rubric, scores: scores) == rubric.criteria.count
    }

    static func statusLabelKey(_ status: PeerReviewAllocationStatus) -> String {
        switch status {
        case .assigned: return "mobile.peerReview.statusAssigned"
        case .inProgress: return "mobile.peerReview.statusInProgress"
        case .submitted: return "mobile.peerReview.statusSubmitted"
        case .expired: return "mobile.peerReview.statusExpired"
        }
    }

    static func cacheKeyAssigned() -> String {
        "peer-review:assigned"
    }

    static func cacheKeyAllocation(_ id: String) -> String {
        "peer-review:allocation:\(id)"
    }

    static func cacheKeyReceived(courseCode: String, assignmentId: String) -> String {
        "peer-review:received:\(courseCode):\(assignmentId)"
    }
}

private extension String {
    var nonEmpty: String? {
        isEmpty ? nil : self
    }
}
